unix/c_editline.c

```diff
@@ -33,6 +33,50 @@ void go_libedit_set_clientdata(EditLine *el, int v) {
     el_set(el, EL_CLIENTDATA, p);
 }
 
+/************* terminals ************/
+
+const char term_me_ansi[] = "\033[0m";
+const char term_bp_off[] = "\033[?2004l";
+const char term_bp_on[] = "\033[?2004h";
+const char term_bp_start[] = "\033[200~";
+const char term_bp_end[] = "\033[201~";
+
+int go_libedit_term_supports_bracketed_paste(EditLine* el) {
+    // termcap has no entry for bracketed paste, so we use the me cap as a
+    // proxy. If the me cap uses the ANSI control code, the terminal will either
+    // support or ignore the bracketed paste ANSI control code.
+    char* me = NULL;
+    if (el_get(el, EL_GETTC, "me", &me, NULL) != 0) {
+	return 0;
+    }
+    return me != NULL && strcmp(me, term_me_ansi) == 0;
+}
+
+static unsigned char _el_bracketed_paste(EditLine *el, int ch) {
+    int endsz = sizeof(term_bp_end) - 1;
+    char* buf = NULL;
+    size_t bufsz = 0;
+    size_t i = 0;
+    while (i < endsz || strncmp(&buf[i-endsz], term_bp_end, endsz) != 0) {
```

Comment:
Wow this is a black magic condition. It's clever (ensuring `strncmp` doesn't get called before `realloc` below is at least called once) but too clever to my taste. Please make this more obvious, and/or add a copious explanatory comment.

Reply:
I'm honestly not sure how to simplify it. Suggestions? The bounds check is not just to avoid accessing `buf` before the call to `realloc`, but to avoid a negative index into buf. That is, even if you initialize `buf` before the loop (`buf = malloc(1024)`), you still need the `i < endsize` condition.